### PR TITLE
refactor(diffusion_planner): move postprocessing logic from node to core

### PR DESCRIPTION
**Pull request by SakodaShintaro**
## Description

Move trajectory, candidate trajectories, predicted objects and turn indicator creation from `publish_predictions` in the node into `DiffusionPlannerCore::create_planner_output`, which returns a `PlannerOutput` struct containing all output messages. This makes the core logic testable independently of the ROS node. The node's `on_timer` now only handles publishing.

## How was this PR tested?

- [x] planning simulator
- [x] logging_simulator

## Notes for reviewers

None.

## Interface changes

None.

## Effects on system behavior

None.

### CODE REVIEW

File: planning/autoware_diffusion_planner/include/autoware/diffusion_planner/diffusion_planner_node.hpp [hunk 1]
```diff
@@ -207,8 +196,6 @@ class DiffusionPlanner : public rclcpp::Node
   VehicleInfo vehicle_info_;
 
   std::unique_ptr<DiagnosticsInterface> diagnostics_inference_;
-  postprocess::TurnIndicatorManager turn_indicator_manager_{
-    rclcpp::Duration::from_seconds(0.0), 0.0f};
 };
```
[the remark this answers is not in the document]
Comment (SakodaShintaro) [reply planning/autoware_diffusion_planner/include/autoware/diffusion_planner/diffusion_planner_node.hpp:198]: https://github.com/autowarefoundation/autoware_universe/pull/12162/commits/8b176d45f9cf7f57d4fb4531c629de7108874a77

File: planning/autoware_diffusion_planner/src/diffusion_planner_core.cpp [hunk 2]
```diff
@@ -284,6 +294,71 @@ TensorrtInference::InferenceResult DiffusionPlannerCore::run_inference(
...
+
+    output.candidate_trajectories.candidate_trajectories.push_back(candidate_trajectory);
+    output.candidate_trajectories.generator_info.push_back(generator_info);
+  }
+
+  // PredictedObjects
+  // Use the first prediction as the main predicted objects
+  constexpr int64_t batch_idx = 0;
+  output.predicted_objects = postprocess::create_predicted_objects(
+    agent_poses, frame_context.ego_centric_neighbor_histories, timestamp, batch_idx);
```
[the remark this answers is not in the document]
Comment (SakodaShintaro) [reply planning/autoware_diffusion_planner/src/diffusion_planner_core.cpp:348]: I will remove `predict_neighbor_trajectory` in another PR

File: planning/autoware_diffusion_planner/src/diffusion_planner_core.cpp [hunk 3]
```diff
@@ -284,6 +294,71 @@ TensorrtInference::InferenceResult DiffusionPlannerCore::run_inference(
...
+
+  // TurnIndicatorsCommand
+  // TODO(sakoda): Use the first logit as the main turn indicator command.
+  // There may be bugs in the current implementation.
```
[the remark this answers is not in the document]
Comment (SakodaShintaro) [reply planning/autoware_diffusion_planner/src/diffusion_planner_core.cpp:352]: I will fix this TOOD in the near future.

File: planning/autoware_diffusion_planner/src/diffusion_planner_core.cpp [hunk 4]
```diff
@@ -284,6 +294,71 @@ TensorrtInference::InferenceResult DiffusionPlannerCore::run_inference(
...
+  const std::vector<float> & predictions, const std::vector<float> & turn_indicator_logit,
+  const FrameContext & frame_context, const rclcpp::Time & timestamp, const UUID & generator_uuid)
+{
+  const auto agent_poses =
+    postprocess::parse_predictions(predictions, frame_context.ego_to_map_transform);
+
+  const bool enable_force_stop =
+    frame_context.ego_kinematic_state.twist.twist.linear.x > std::numeric_limits<double>::epsilon();
+
+  PlannerOutput output;
```
[the remark this answers is not in the document]
Comment (SakodaShintaro) [reply planning/autoware_diffusion_planner/src/diffusion_planner_core.cpp:306]: I may add a total test in another repository in the future.

File: planning/autoware_diffusion_planner/src/diffusion_planner_core.cpp [hunk 5]
```diff
@@ -284,6 +294,71 @@ TensorrtInference::InferenceResult DiffusionPlannerCore::run_inference(
...
+
+    const auto candidate_trajectory = autoware_internal_planning_msgs::build<
+                                        autoware_internal_planning_msgs::msg::CandidateTrajectory>()
+                                        .header(trajectory.header)
+                                        .generator_id(generator_uuid)
+                                        .points(trajectory.points);
+
+    std_msgs::msg::String generator_name_msg;
+    generator_name_msg.data = "DiffusionPlanner_batch_" + std::to_string(i);
+
```
[the remark this answers is not in the document]
Comment (SakodaShintaro) [reply planning/autoware_diffusion_planner/src/diffusion_planner_core.cpp:334]: Let's see if clang-tidy points this out.

File: planning/autoware_diffusion_planner/src/diffusion_planner_node.cpp [hunk 6]
```diff
@@ -27,8 +27,6 @@
 #include <rclcpp/duration.hpp>
 #include <rclcpp/logging.hpp>
 
-#include <autoware_internal_planning_msgs/msg/candidate_trajectory.hpp>
-#include <autoware_internal_planning_msgs/msg/generator_info.hpp>
 #include <autoware_perception_msgs/msg/tracked_objects.hpp>
 
 #include <Eigen/src/Core/Matrix.h>
```
[the remark this answers is not in the document]
Comment (SakodaShintaro) [reply planning/autoware_diffusion_planner/src/diffusion_planner_node.cpp]: https://github.com/autowarefoundation/autoware_universe/pull/12162/commits/091d8b5ad5e7b389342b04f0bb0b4a767e95f5cb

File: planning/autoware_diffusion_planner/src/diffusion_planner_node.cpp [hunk 7]
```diff
@@ -417,13 +331,15 @@ void DiffusionPlanner::on_timer()
   }
   const auto & [predictions, turn_indicator_logit] = inference_result.outputs.value();
 
-  publish_predictions(predictions, *frame_context, frame_time);
+  const auto planner_output = core_->create_planner_output(
```
Comment (go-sakayori) [inline planning/autoware_diffusion_planner/src/diffusion_planner_node.cpp]: I suggest to wrap `core_->create_planner_output(...)`  with `try/catch` and report `ERROR` diagnostics on failure, so exceptions from postprocessing don’t break the timer loop. Any thoughts?

File: planning/autoware_diffusion_planner/src/diffusion_planner_core.cpp (hunk 5 above)
[the remark this answers is not in the document]
Comment (yukkysaito) [reply planning/autoware_diffusion_planner/src/diffusion_planner_core.cpp:334]: This works, but I'd recommend making the string construction explicit:

```
generator_name_msg.data =
  std::string("DiffusionPlanner_batch_") + std::to_string(i);
```

This avoids relying on the const char* + std::string overload and makes the intent clearer.

File: planning/autoware_diffusion_planner/src/diffusion_planner_core.cpp [hunk 8]
```diff
@@ -284,6 +293,71 @@ TensorrtInference::InferenceResult DiffusionPlannerCore::run_inference(
...
+  for (int i = 0; i < params_.batch_size; i++) {
+    auto trajectory = postprocess::create_ego_trajectory(
+      agent_poses, timestamp, frame_context.ego_kinematic_state.pose.pose.position, i,
+      params_.velocity_smoothing_window, enable_force_stop, params_.stopping_threshold);
+
+    if (params_.shift_x) {
+      for (auto & point : trajectory.points) {
+        point.pose = utils::shift_x(point.pose, -vehicle_info_.wheel_base_m / 2.0);
+      }
+    }
```
Comment (yukkysaito) [inline planning/autoware_diffusion_planner/src/diffusion_planner_core.cpp:319]: Although this is slightly outside the scope of this refactoring, I have a question regarding shift_x.

Is the purpose of shift_x to compute the vehicle center position from base_link?

If so, in order to compute it accurately, we would also need to take both front_overhang and rear_overhang into account.

File: planning/autoware_diffusion_planner/src/diffusion_planner_core.cpp [hunk 9]
```diff
@@ -284,6 +293,71 @@ TensorrtInference::InferenceResult DiffusionPlannerCore::run_inference(
...
+  }
+
+  // PredictedObjects
+  // Use the first prediction as the main predicted objects
+  constexpr int64_t batch_idx = 0;
+  output.predicted_objects = postprocess::create_predicted_objects(
+    agent_poses, frame_context.ego_centric_neighbor_histories, timestamp, batch_idx);
+
+  // TurnIndicatorsCommand
+  // TODO(sakoda): Use the first logit as the main turn indicator command.
```
Comment (yukkysaito) [inline planning/autoware_diffusion_planner/src/diffusion_planner_core.cpp:351]: Does this mean that the multi-batch handling of the turn indicator is not being handled properly?

Comment (SakodaShintaro) [reply planning/autoware_diffusion_planner/src/diffusion_planner_core.cpp:351]: Yes, I think that the current implementation uses the max operator over the entire logits, not just batch 0.

File: planning/autoware_diffusion_planner/src/diffusion_planner_core.cpp (hunk 8 above)
Comment (SakodaShintaro) [reply planning/autoware_diffusion_planner/src/diffusion_planner_core.cpp:319]: I mistakenly thought that base_link alone was sufficient (assuming front_overhang == rear_overhang).
I added a comment.
https://github.com/autowarefoundation/autoware_universe/pull/12162/commits/b6a23c245345fbb4c1ea81e9f1f67ce65f3526f0

File: planning/autoware_diffusion_planner/src/diffusion_planner_node.cpp (hunk 7 above)
Comment (SakodaShintaro) [reply planning/autoware_diffusion_planner/src/diffusion_planner_node.cpp]: https://github.com/autowarefoundation/autoware_universe/pull/12162/commits/18833a7f6b3cb99baaed61d985a23ea1e9068ce3

File: planning/autoware_diffusion_planner/src/diffusion_planner_core.cpp (hunk 5 above)
[the remark this answers is not in the document]
Comment (SakodaShintaro) [reply planning/autoware_diffusion_planner/src/diffusion_planner_core.cpp:334]: Fixed in https://github.com/autowarefoundation/autoware_universe/pull/12162/commits/939f5f3db78180f542a76c7baf96bb58bc8a8598

**Review by yukkysaito — APPROVED** [review]
